### PR TITLE
refactor to add jwt cache and authenticator object

### PR DESCRIPTION
**Pull request by techcobweb**
Signed-off-by: Mike Cobbett <77053+techcobweb@users.noreply.github.com>

# Why ?
This is part of this issue: Local test run to be able to use remote CPS
[#1813](https://github.com/galasa-dev/projectmanagement/issues/1813)

# Tasks in this PR
- [x] Add a 'JWT cache' concept inside the authenticator, preparing the way to encrypt the data.
- [x] Add an authenticator class to pull together lots of tightly-bound functions
- ~~Logoff should accept a bootstrap flag to make it symmetrical~~ abandoned because we want to move to a context and away from bootstrap
- [x] Logoff should remove the file using cache.clear rather than zapping the underlying file.
- [x] Move some interfaces to an `spi` folder so we can avoid cyclic dependencies easier.
- [x] The bearer token code in utils is the only thing which knows where bearer tokens are stored. (In the `~/.galasa/bearer-tokens` folder)
- [x] The cache is the only thing which manipulates bearer tokens
- [x] Logout flushes all cache entries
- [x] A JWT in the cache which has expired is treated as not existing at all, so a new one can be allocated.
- [x] Linted the code a bit, and cleaned up some of the warnings.
- [x] Each server has it's own jwt stored in the cache. Stored in a folder whose name is the uuencoded api server URL. So they are unique files.
- [x] The files are encrypted before they are written to disk. Using the galasa token.
- [ ] The files are accessed using exclusive lock access. So one tool instance can't interfere with another.

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

**Review by eamansour — APPROVED** [review]

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-cli-pr-240-9wq8v for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: pkg/auth/authLogin_test.go [hunk 1]
```diff
@@ -98,126 +111,136 @@ func TestLoginCreatesBearerTokenFileContainingJWT(t *testing.T) {
...
+// 	err := authenticator.Login()
 
-	// Then...
-	assert.NotNil(t, err, "Should return an error if the GALASA_ACCESS_TOKEN property is missing")
-	assert.ErrorContains(t, err, "GAL1122E")
-}
+// 	// Then...
+// 	assert.NotNil(t, err, "Should return an error if the GALASA_ACCESS_TOKEN property is missing")
+// 	assert.ErrorContains(t, err, "GAL1122E")
+// }
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/auth/authLogin_test.go:223]: yep

File: pkg/errors/errorMessage.go [hunk 2]
```diff
@@ -234,6 +234,8 @@ var (
 		"The %s field value must start with a character in the 'a-z' or 'A-Z' range, followed by any characters in the 'a'-'z', 'A'-'Z', '0'-'9', '.' (period), '-' (dash) or '_' (underscore) ranges only.", 1142, STACK_TRACE_NOT_WANTED)
 	GALASA_ERROR_QUERY_RUNS_NON_OK_STATUS         = NewMessageType("GAL1143E: Could not query run results. Server returned a non-200 code (%s)", 1143, STACK_TRACE_NOT_WANTED)
 	GALASA_ERROR_GET_TEST_CATALOG_CONTENTS_FAILED = NewMessageType("GAL1144E: Could not use url '%s' to retrieve the contents of the test catalog from stream '%s'. Http error from the Galasa server is '%v'", 1144, STACK_TRACE_NOT_WANTED)
+	GALASA_JWT_CANNOT_BE_PARSED                   = NewMessageType("GAL1145E: Jwt token is invalid or can't be parsed: %s", 1145, STACK_TRACE_NOT_WANTED)
+	GALASA_JWT_HAS_NO_EXPIRATION_DATETIME         = NewMessageType("GAL1146E: Failed to get the expiration time from the jwt claim: %s\n", 1146, STACK_TRACE_NOT_WANTED)
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/errors/errorMessage.go:238]: Indeed.

File: pkg/auth/authLogin_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/auth/authLogin_test.go:223]: Fixed

**Comment by galasa-team** [comment]
Build successful

File: pkg/errors/errorMessage.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/errors/errorMessage.go]: Re-worded with more detail and advice.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: pkg/auth/authLogin_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (techcobweb) [reply pkg/auth/authLogin_test.go]: Re-instated.

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-cli-pr-240-lbczh for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-cli-pr-240-c8c86 for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**